frontend/catalyst/utils/runtime.py

```diff
@@ -159,77 +170,125 @@ def is_not_ctrl(op):
 
     operations_no_adj = filter(is_not_adj, operations)
     operations_no_adj_no_ctrl = filter(is_not_ctrl, operations_no_adj)
-    return list(operations_no_adj_no_ctrl)
+    return set(operations_no_adj_no_ctrl)
 
 
-def check_full_overlap(device, *args):
+def check_full_overlap(device_gates: Set[str], spec_gates: Set[str]) -> None:
     """Check that device.operations is equivalent to the union of *args
 
     Args:
-        device (qml.Device): An instance of a quantum device.
-        *args (List[Str]): List of strings.
+        device_gates (Set[str]): device gates
+        spec_gates (Set[str]): spec gates
 
     Raises: CompileError
     """
-    operations = filter_out_adjoint_and_control(device.operations)
-    gates_in_device = set(operations)
-    set_of_sets = [set(arg) for arg in args]
-    union = set.union(*set_of_sets)
-    if gates_in_device == union:
+    if device_gates == spec_gates:
         return
 
-    msg = "Gates in qml.device.operations and specification file do not match"
+    msg = (
+        "Gates in qml.device.operations and specification file do not match.\n"
+        f"Gates that present only in the device: {device_gates - spec_gates}\n"
+        f"Gates that present only in spec: {spec_gates - device_gates}\n"
+    )
     raise CompileError(msg)
 
 
-def check_gates_are_compatible_with_device(device, config):
-    """Validate configuration dictionary against device.
+def validate_config_with_device(device: qml.QubitDevice, config: TOMLDocument) -> None:
+    """Validate configuration document against the device attributes.
+    Raise CompileError in case of mismatch.
+    * if configuration file does not exists
+    * if decomposable, matrix, and native gates have some overlap
+    * if decomposable, matrix, and native gates do not match gates in ``device.operations``
 
     Args:
         device (qml.Device): An instance of a quantum device.
-        config (Dict[Str, Any]): Configuration dictionary
 
     Raises: CompileError
     """
 
-    native = get_native_gates(config)
-    decomposable = get_decomposable_gates(config)
-    matrix = get_matrix_decomposable_gates(config)
-    check_no_overlap(native, decomposable, matrix)
-    if not hasattr(device, "operations"):  # pragma: nocover
-        # The new device API has no "operations" field
-        # so we cannot check that there's an overlap or not.
-        return
+    if not config["compilation"]["qjit_compatible"]:
+        raise CompileError(
+            f"Attempting to compile program for incompatible device '{device.name}': "
+            f"Config is not marked as qjit-compatible"
+        )
 
-    check_full_overlap(device, native, decomposable, matrix)
+    shots_present = device.shots is not None
+    native = get_pennylane_operations(config, shots_present)
```

Comment:
It is because it returns strings in the Pennylane's format, i.e. with `C(Gate)` syntax. Calculating it requires some additional logic, please check the sources.

Reply:
I would actually like us to move away from the `C()` syntax entirely. 
Maybe there are some things to consider (like schema v1 or old device API), but what would like us to do is determine compatibility with an operation as follows:
- is gate in native gates -> good
- is gate controlled
   - is base gate in native gates with controllable attribute -> good
- is gate adjoint
   - is base gate in native gates with invertible attribute -> good

Reply:
Its not our syntax, it is PennyLane's! It uses it when it check the decomposition criteria! This function just translates the toml config into what PL understands.